src/lib.rs

```diff
@@ -87,46 +98,80 @@ pub fn get_zones_and_stats_without_geom(
     Ok((zones, stats))
 }
 
-fn get_country<'a>(_zone: &zone::Zone, country_code: &'a Option<String>) -> Result<&'a str, Error> {
+fn get_country_code<'a>(
+    country_finder: &'a CountryFinder,
+    zone: &zone::Zone,
+    country_code: &'a Option<String>,
+) -> Option<String> {
     if let &Some(ref c) = country_code {
-        Ok(c)
+        Some(c.clone())
     } else {
-        //TODO add a realway to find the country
-        Err(failure::err_msg("Cannot find the country of the zone"))
+        country_finder.find_zone_country(&zone)
     }
 }
 
-fn create_ontology(
+fn type_zones(
     zones: &mut Vec<zone::Zone>,
     stats: &mut CosmogonyStats,
     libpostal_file_path: PathBuf,
     country_code: Option<String>,
 ) -> Result<(), Error> {
     let zone_typer = zone_typer::ZoneTyper::new(libpostal_file_path)?;
+    let country_finder: CountryFinder = zones.iter().collect();
+    if country_code.is_none() && country_finder.is_empty() {
+        return Err(failure::err_msg(
+            "no country_code has been provided and no country have been found, we won't be able to make a cosmogony",
+        ));
+    }
 
     for mut z in zones {
-        let country = get_country(&z, &country_code)?;
-        let type_res = zone_typer.get_zone_type(&z, &country);
-        match type_res {
-            Ok(t) => z.zone_type = Some(t),
-            Err(zone_typer::ZoneTyperError::InvalidCountry(c)) => {
-                info!("impossible to find {}", c);
-                *stats.zone_with_unkwown_country.entry(c).or_insert(0) += 1;
+        let country_code = get_country_code(&country_finder, &z, &country_code);
```

Comment:
I fear that there might be an uppercase / lowercase difference between the code returned by the country_finder and the libpostal filenames. Are we sure that's handled correctly ?

Reply:
it is done while reading the countrycode, but you're right I moved it in the libpostal lookup, it's more consistent